argo-workflows/nautobot-interface-sync/code/models.py

```diff
@@ -0,0 +1,114 @@
+from dataclasses import dataclass
+
+
+@dataclass
+class Interface:
+    name: str
+    mac_addr: str
+    location: str
+    current_speed_mbps: int
+    nic_model: str
+
+    @classmethod
+    def from_redfish(cls, data, nic):
+        if data.root.json["Vendor"] == "HPE":
+            name = f"{nic.name}_{data.physical_port_number}"
+        else:
+            name = data.identity
+        return cls(
+            name,
+            data.associated_network_addresses[0],
+            nic.location,
+            data.current_link_speed_mbps,
+            nic.model,
+        )
+
+    @classmethod
+    def from_hp_json(cls, data, nic, ports):
+        p_num = data.get("port_num") or (ports.index(data) + 1)
+        intfs_name = f"NIC.{nic.location.replace(' ', '.')}_{str(p_num)}"
+        return cls(
+            intfs_name,
+            data.get("mac_addr"),
+            nic.location,
+            data.get("speed", 0),
+            nic.model,
+        )
+
+
+@dataclass
+class NIC:
+    name: str
+    location: str
+    interfaces: list
+    model: str
+
+    @classmethod
+    def from_redfish(cls, data):
+        location = cls.nic_location(data)
+        nic = cls(data.identity, location, [], data.model)
+        nic.interfaces = [Interface.from_redfish(i, nic) for i in cls.nic_ports(data)]
+        return nic
+
+    @classmethod
+    def from_hp_json(cls, data):
+        nic = cls(data.get("name"), data.get("location"), [], data.get("name"))
+        ports = data.get("network_ports") or data.get("unknown_ports")
+        nic.interfaces = [Interface.from_hp_json(i, nic, ports) for i in ports]
+        return nic
+
+    @classmethod
+    def nic_location(cls, nic):
+        try:
+            return nic.json["Controllers"][0]["Location"]["PartLocation"][
+                "ServiceLabel"
+            ]
+        except KeyError:
+            return nic.identity
+
+    @classmethod
+    def nic_ports(cls, nic):
+        return nic.network_ports.get_members()
+
+
+@dataclass
+class Chassis:
+    name: str
+    nics: list
+    network_interfaces: list
+
+    @classmethod
+    def from_redfish(cls, oob_obj):
```

Comment:
What do you think about adding a pre-flight sort of check to see if the chassis is supported?
For example, if we suddenly stumbled upon a server manufactured by Xiaomi, should we bail with an error or still try to discover and potentially produce invalid data?

Reply:
added